Add `Reject` handling for Application user follow requests

pfefferle opened the pull request:
Prevents application users from being followed by automatically sending a 'Reject' response when a follow request targets the application user.

<!--- Provide a general summary of your changes in the Title above -->

For reference: https://github.com/Automattic/wordpress-activitypub/issues/2099#issuecomment-3228716657

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds automatic rejection of follow requests for application users
* Introduces a new queue_reject method to handle rejection responses


### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Try to follow the Application Actor
* You should receive a `Reject`

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [X] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [X] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Prevents Application users from being followed.

</details>

pfefferle committed:
Add reject handling for application user follow requests

Prevents application users from being followed by automatically sending a 'Reject' response when a follow request targets the application user. Also fixes the recipient of the 'Accept' activity to use the correct actor.

pfefferle committed:
Fix variable overwrite in queue_reject method

Renamed the filtered activity array to $origin_activity to prevent overwriting the $activity object. Updated references to use $origin_activity when setting the object and recipients for the 'Reject' activity.

pfefferle committed:
Fix recipient in follow accept activity

Replaces the recipient of the 'Accept' activity from the object's actor to the correct actor variable, ensuring the activity is sent to the intended recipient.

matticbot committed:
Add changelog

pfefferle committed:
Remove pending follow requests on migration

Adds a migration step to delete pending follow requests for the application user when upgrading from versions earlier than 'unreleased'. Introduces the remove_pending_application_user_follow_requests() method to handle this cleanup.

pfefferle committed:
Add tests for removing pending application user follow requests

Introduces unit tests for Migration::remove_pending_application_user_follow_requests to verify correct removal of _activitypub_following post meta entries with APPLICATION_USER_ID. Tests cover scenarios with matching, non-matching, and multiple entries to ensure only the intended meta entries are deleted and others remain unaffected.

pfefferle committed:
Add phpcs:ignore for direct DB queries in migration tests

Added phpcs:ignore WordPress.DB.DirectDatabaseQuery comments to all direct $wpdb->get_var() calls in class-test-migration.php to suppress coding standard warnings. Also updated some queries to use $wpdb->prepare for parameterization and consistency.

pfefferle committed:
Add tests for handle_follow method in Test_Follow

Introduces a new test for the handle_follow method to verify correct handling of 'Accept' and 'Reject' activity types in the outbox. Also updates get_posts calls to use the global namespace for consistency.

pfefferle committed:
Update application summary to static message with domain

Replaces the dynamic blog description with a static, translatable message that includes the site's domain. This clarifies the purpose of the Application Actor and discourages user interaction.

obenland commented on includes/handler/class-follow.php:
```diff
@@ -100,4 +105,33 @@ public static function queue_accept( $actor, $activity_object, $user_id, $remote
+		// Only send minimal data.
+		$origin_activity = array_intersect_key(
+			$activity,
+			array_flip(
+				array(
+					'id',
+					'type',
+					'actor',
+					'object',
+				)
+			)
```
```suggestion
			array(
				'id'     => 1,
				'type'.  => 1,
				'actor'  => 1,
				'object' => 1,
			)
```

pfefferle replied:
???

obenland replied:
we can save the array_flip call by defining the array with the right key value pairs

pfefferle replied:
I see! Context wins :)

obenland commented on includes/model/class-application.php:
```diff
@@ -241,11 +241,10 @@ public function get_public_key() {
-			)
+		return sprintf(
+			/* translators: %s: Domain of the site */
+			__( 'This is the Application Actor for %s. It&#8217;s a tool account only, please don&#8217;t follow or interact with it.', 'activitypub' ),
```
Feels odd to say that, no? Maybe just return an empty summary instead?

pfefferle replied:
Why? This user should simply not be followed at all.

obenland replied:
"please don't follow or interact with it" just sounds odd. It shouldn't be up to others, we shouldn't put the onus on others to do or not do something. If anything I bet it'll make people want to interact with it more :)

pfefferle replied:
the problem is, that people do, becaue I have follow requests for that user. we can remove the please do not follow and only add the first part.

obenland commented on includes/class-migration.php:
```diff
@@ -1055,4 +1059,19 @@ public static function update_actor_json_storage( $batch_size = 100 ) {
+		global $wpdb;
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery
+		$wpdb->query(
+			$wpdb->prepare(
+				"DELETE FROM {$wpdb->postmeta} WHERE meta_key = '_activitypub_following' AND meta_value = %s",
+				Actors::APPLICATION_USER_ID
+			)
+		);
```
Would this be a candidate for `$wpdb->delete()`?

pfefferle replied:
Fixed in https://github.com/Automattic/wordpress-activitypub/pull/2101/commits/84d362e3437bb56fcb74bec3a67ba3a5e1a829ef

pfefferle committed:
Refactor postmeta deletion to use $wpdb->delete

Replaces a direct SQL DELETE query with the $wpdb->delete method for removing '_activitypub_following' meta entries. This improves code readability and leverages WordPress's built-in database abstraction.

pfefferle committed:
Simplify array_intersect_key usage in Follow handler

Replaces array_flip(array(...)) with direct associative arrays in array_intersect_key calls for improved readability and minor performance benefit.

pfefferle committed:
Simplify application actor summary message

Removed the note discouraging following or interacting with the application actor from the summary message in class-application.php.

pfefferle committed:
Use home_host() in Application actor description

Replaces direct call to wp_parse_url(home_url(), PHP_URL_HOST) with the home_host() function for retrieving the site's domain in the Application actor description.
